cpp/src_prims/metrics/trustworthiness_score.cuh

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <rmm/device_uvector.hpp>
+
+#include <raft/distance/distance.cuh>
+#include <raft/spatial/knn/knn.hpp>
+#include <selection/columnWiseSort.cuh>
+
+#define N_THREADS 512
+
+namespace MLCommon {
+namespace Score {
+
+/**
+     * @brief Compute a the rank of trustworthiness score
+     * @param[in] X_ind: Indexes of pairwise distance calculations of X and sorting
+     * @param[in] X_dist: Distances of pairwise distance calculations of X and sorting
+     * @param[in] emb_ind: Indexes of KNN on embeddings
+     * @param[in] emb_dist: Distances of KNN on embeddings
+     * @param n: Number of samples
+     * @param n_neighbors: Number of neighbors considered by trustworthiness score
+     * @param work: Batch to consider (to do it at once use n * n_neighbors)
+     * @param[out] rank: Resulting rank
+     */
+template <typename math_t, typename knn_index_t>
+__global__ void compute_rank(int *X_ind, math_t *X_dist, knn_index_t *emb_ind,
+                             math_t *emb_dist, int n, int n_neighbors, int work,
+                             int offset, double *rank) {
+  int i = blockIdx.x * blockDim.x + threadIdx.x;
+  if (i >= work) return;
+
+  int sample_idx = i / n_neighbors;
+  int nn_idx = i % n_neighbors;
+
+  int loc = sample_idx * n_neighbors + nn_idx;
+  knn_index_t emb_nn_ind = emb_ind[loc];
+
+  if ((offset + nn_idx) == emb_nn_ind) return;
+
+  math_t emb_nn_dist = emb_dist[loc];
+
+  loc = sample_idx * n;
+  int *X_pw_ind = &X_ind[loc];
+  math_t *X_pw_dist = &X_dist[loc];
+
+  knn_index_t left = 0;
+  knn_index_t right = n - 1;
+  while (left < right) {
+    knn_index_t mid = std::floor((left + right) / 2);
+    if (X_pw_dist[mid] < emb_nn_dist) {
+      left = mid + 1;
+    } else {
+      right = mid;
+    }
+  }
+
+  for (int r = left; r < n; r++) {
+    if (X_pw_ind[r] == emb_nn_ind) {
+      int tmp = r - n_neighbors + 1;
+      if (tmp > 0) raft::myAtomicAdd<double>(rank, tmp);
+      break;
+    }
+  }
+}
```

Comment:
After taking a look at the speedup, it looks like the simple loop versoin is 25% faster than the binary search on a high quality embedding. The runtimes look similar with a somewhat degraded embedding. I think that the main weaknesses of the binary search solution is that accesses are not coalesced and that the first few access are similar with almost every thread. @cjnolet should we opt for a simple loop instead, or maybe is there a better solution?

Reply:
@viclafargue, upon closer look at your rank kernel changes, I believe you are making an implicit assumption that the relative distances in the embedded space should correspond directly to the distances in the original space. Is it possible the tests passed by coincidence here?

Reply:
You're right, that's a mistake. Thanks for noticing, I'll fix this.